### PR TITLE
Use ByteString for GroupEntry/UserEntry

### LINKED ISSUE #237: `GroupEntry`/`UserEntry` uses String (even in ByteString variants)
https://github.com/haskell/unix/issues/63#issuecomment-1179805976

String has encoding issues. These things should be NUL terminated byte arrays that we get from the C API. The problem is how to unify across String and ByteString modules, since these are types, not just functions:

1. use ByteString based record for both APIs
2. keep String based records
3. provide two different types
4. parametrize the type with the String type
5. don't expose the constructor/records, but provide "getters" (these can do the `ByteString -> String` conversion in the String API)

### PR DESCRIPTION
**Pull request by hasufell**
No PosixString instance for now. Not sure
if it's interesting.

Fixes #237 

@Bodigrim

### CODE REVIEW

File: System/Posix/User.hsc [hunk 1]
```diff
@@ -371,13 +377,30 @@ getEffectiveUserName = do
...
+-- | Manually constructing 'GroupEntry' in String modules is discouraged. It will truncate
+-- Chars to 8bit. Use 'System.Posix.User.ByteString' instead.
+pattern GroupEntry :: String          -- ^ The name of this group (gr_name)
+                   -> String          -- ^ The password for this group (gr_passwd)
+                   -> GroupID         -- ^ The unique numeric ID for this group (gr_gid)
+                   -> [String]        -- ^ A list of zero or more usernames that are members (gr_mem)
+                   -> GroupEntry
+pattern GroupEntry gn gp gi gm <- MkGroupEntry (C8.unpack -> gn) (C8.unpack -> gp) gi (fmap C8.unpack -> gm) where
+  GroupEntry gn gp gi gm = MkGroupEntry (C8.pack gn) (C8.pack gp) gi (C8.pack <$> gm)
+#if __GLASGOW_HASKELL__ >= 802
```
Comment (Bodigrim) [inline System/Posix/User.hsc:401]: We do not support GHC 8.0 anyway.

File: System/Posix/User/Common.hsc [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+   gecos  <- (#peek struct passwd, pw_gecos)  ptr >>= packCString
+#endif
+   dir    <- (#peek struct passwd, pw_dir)    ptr >>= packCString
+   shell  <- (#peek struct passwd, pw_shell)  ptr >>= packCString
+   return (MkUserEntry name passwd uid gid gecos dir shell)
+
+#endif // HAVE_PWD_H
+
+data UserEntry =
+ MkUserEntry {
```
Comment (Bodigrim) [inline System/Posix/User/Common.hsc:63]:
```suggestion
 UserEntry {
```
And re-export it directly from `ByteString.hsc`.

**Review by Bodigrim — APPROVED** [review]

**Comment by Bodigrim** [comment]
@hs-viktor @vdukhovny could you please take a look?

**Review by vdukhovni — COMMENTED** [review]
Mostly LGTM.

My only concern is that the GECOS field may well be UTF8 encoded on some systems.  It is human-edited/readable data and in various non-US locales could well hold strings in that locale.  Of course that locale's LC_CTYPE need not be UTF8.

At the very least we should call out the change.  Users of the `String` API might not get the string they expect or were used to.  They could switch to the ByteString API and apply the relevant encoding.  Or we could consider attempting to apply the locale's character set to the GECOS field, though of course the user's locale may not match the system-administrator's locale (and then various non-Haskell utilities would also show garbage for non-ASCII gecos fields).

Perhaps try to decode in the current locale, and if that fails, just unpack the ByteString?

Or would it be better to just always unpack (current code) and leave the caller to repack and decode in some suitable encoding???

**Comment by hasufell** [comment]
> Mostly LGTM.
> 
> My only concern is that the GECOS field may well be UTF8 encoded on some systems.  It is human-edited/readable data and in various non-US locales could well hold strings in that locale.  Of course that locale's LC_CTYPE need not be UTF8.
> 
> At the very least we should call out the change.  Users of the `String` API might not get the string they expect or were used to.  They could switch to the ByteString API and apply the relevant encoding.  Or we could consider attempting to apply the locale's character set to the GECOS field, though of course the user's locale may not match the system-administrator's locale (and then various non-Haskell utilities would also show garbage for non-ASCII gecos fields).
> 
> Perhaps try to decode in the current locale, and if that fails, just unpack the ByteString?
> 
> Or would it be better to just always unpack (current code) and leave the caller to repack and decode in some suitable encoding???

`getpwnam` is not POSIX. Any manpage I looked at doesn't specify an encoding. It's NUL terminated char pointer.

Whether most systems use UTF-something is irrelevant. Guessing the encoding based on current locale is wrong.

The current string implementation assumes Latin-1 and doesn't do unicode decoding either. This PR maintains that behavior for the String module. I don't think changing that makes it any more or less correct. Both are wrong.

Only the ByteString variant is correct.

**Comment by vdukhovni** [comment]
Since the original (pre-PR) code is equivalent to just an `unpack` of the GECOS, then indeed we're not really changing anything. I had assumed that it previously applied a locale-specific decoder, but that's not the case.

**Review by vdukhovni — APPROVED** [review]

**Pull request merged**